astlib/location.ml

```diff
@@ -112,4 +124,6 @@ module Error = struct
     (*IF_AT_LEAST 408 _set_main_loc_new error loc*)
 end
 
-let raise_errorf ?loc msg = raise_errorf ?loc msg
+let raise_errorf ?loc msg =
+  (*IF_AT_LEAST 503 Format.kasprintf (fun s -> raise_errorf ?loc "%s" s) msg *)
```

Comment:
This part looks problematic: you are rendering to a string too early. 
I can see two options:
- either update the type of `raise_errorf` to the new printers from `Format_doc`
- or use the deprecated escape-hatch from `Format_doc`:
```ocaml
Format.kdprintf (fun pr -> raise_errorf ?loc "%t" (Format_doc.deprecated_printer pr)) msg
```

Reply:
Note that the escape hatch is here to stay, but it will impede serialization of error reports in the future.

Reply:
Ah thanks! I'll update to you suggestion. We unfortunately cannot update the type of raise_errorf since we both need to stay compatible with a range of OCaml versions in which `Format_doc` does not exist and the type change would be a breaking change (at least for some users, depending on how `raise_errorf` is used I expect in some cases it would be transparent).

We'll have to support it's current form for a little while.

Reply:
It also seems that we'll need to bump our lower bound to `4.08` to use `Format.kdprintf`. This is something we could consider doing though.

Reply:
It is a bit cumbersome but ppxlib could also vendor the definition of `kdprintf`:
```
open CamlinternalFormat
open Format
open CamlinternalFormatBasics

let compute_tag output tag_acc =
  let buf = Buffer.create 16 in
  let ppf = formatter_of_buffer buf in
  output ppf tag_acc;
  pp_print_flush ppf ();
  let len = Buffer.length buf in
  if len < 2 then Buffer.contents buf
  else Buffer.sub buf 1 (len - 2)
let pp_open_box_gen ppf indent box_type =
  match box_type with
  | Pp_vbox -> pp_open_vbox ppf indent
  | Pp_hbox | Pp_fits -> pp_open_hbox ppf ()
  | Pp_hovbox -> pp_open_hovbox ppf indent
  | Pp_hvbox -> pp_open_hvbox ppf indent
  | Pp_box -> pp_open_box ppf indent
 (* Interpret a formatting entity on a formatter. *)
let output_formatting_lit ppf fmting_lit = match fmting_lit with
  | Close_box                 -> pp_close_box ppf ()
  | Close_tag                 -> pp_close_stag ppf ()
  | Break (_, width, offset)  -> pp_print_break ppf width offset
  | FFlush                    -> pp_print_flush ppf ()
  | Force_newline             -> pp_force_newline ppf ()
  | Flush_newline             -> pp_print_newline ppf ()
  | Magic_size (_, _)         -> ()
  | Escaped_at                -> pp_print_char ppf '@'
  | Escaped_percent           -> pp_print_char ppf '%'
  | Scan_indic c              -> pp_print_char ppf '@'; pp_print_char ppf c
let rec output_acc ppf acc = match acc with
  | Acc_string_literal (Acc_formatting_lit (p, Magic_size (_, size)), s)
  | Acc_data_string (Acc_formatting_lit (p, Magic_size (_, size)), s) ->
    output_acc ppf p;
    pp_print_as ppf size s;
  | Acc_char_literal (Acc_formatting_lit (p, Magic_size (_, size)), c)
  | Acc_data_char (Acc_formatting_lit (p, Magic_size (_, size)), c) ->
    output_acc ppf p;
    pp_print_as ppf size (String.make 1 c);
  | Acc_formatting_lit (p, f) ->
    output_acc ppf p;
    output_formatting_lit ppf f;
  | Acc_formatting_gen (p, Acc_open_tag acc') ->
    output_acc ppf p;
    pp_open_stag ppf (String_tag (compute_tag output_acc acc'))
  | Acc_formatting_gen (p, Acc_open_box acc') ->
    output_acc ppf p;
    let (indent, bty) = open_box_of_string (compute_tag output_acc acc') in
    pp_open_box_gen ppf indent bty
  | Acc_string_literal (p, s)
  | Acc_data_string (p, s)   -> output_acc ppf p; pp_print_string ppf s;
  | Acc_char_literal (p, c)
  | Acc_data_char (p, c)     -> output_acc ppf p; pp_print_char ppf c;
  | Acc_delay (p, f)         -> output_acc ppf p; f ppf;
  | Acc_flush p              -> output_acc ppf p; pp_print_flush ppf ();
  | Acc_invalid_arg (p, msg) -> output_acc ppf p; invalid_arg msg;
  | End_of_acc               -> ()

let kdprintf k (Format (fmt, _)) =
  make_printf
    (fun acc -> k (fun ppf -> output_acc ppf acc))
    End_of_acc fmt
 ```

Reply:
I think we can actually consider requiring 4.08+. In the meantime I suggest we merge the 5.3 support as it is and we have time to improve this later on!

If we do end up choosing to require 4.08+ we can then simply use your suggested implementation, using `kdprintf`!